### PR TITLE
Tidy up a few things in the readme and code.

### PR DESCRIPTION
**Pull request by PromoFaux**
### **What does this PR aim to accomplish?:**

Dipping my toes back into the water. 

 - Tidy up a few things in the Readme file
   - Add some more comments to the example compose file
   - Add that setting the container's DNS to be itself is not recommended 
 - Remove some commended out code from startup script
 - Move the webpassword checking section into `ensure_basic_configuration`
 - Ensure that FTL's `dns.upstreams` config is set (either in environment or config file) hard exit if not.

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code and I have tested my changes.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

---
- [x] I have read the above and my PR is ready for review. _Check this box to confirm_

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -25,31 +25,43 @@ Using [Docker-compose](https://docs.docker.com/compose/install/):
...
+    image: pihole/pihole:latest    
     ports:
+      # DNS Ports
       - "53:53/tcp"
       - "53:53/udp"
-      - "67:67/udp" # Only required if you are using Pi-hole as your DHCP server
+      # Default HTTP Port
       - "80:80/tcp"
-      - "443:443/tcp" # By default, FTL will generate a self-signed certificate
+      # Default HTTPs Port. FTL Will generate a self-signed certificate
```
Comment (yubiuser) [inline README.md:44]:
```suggestion
      # Default HTTPs Port. FTL will generate a self-signed certificate
```

**Review by yubiuser — CHANGES_REQUESTED** [review]
Test need adjustment.

https://github.com/pi-hole/docker-pi-hole/blob/2cabcc6fd1ce27e4586cf41d20e25c0e8a7e8c51/test/tests/test_bash_functions.py#L18

File: README.md [hunk 2]
```diff
@@ -72,7 +84,7 @@ It is recommended that you use environment variables to configure the Pi-hole do
 
 ### Web interface password
 
-To set a specific password for the web interface, use the environment variable `FTLCONF_webserver_api_password`. If this variable is not detected, and you have not already set one via `pihole setpassword` in the container, then a random password will be assigned on startup, this will be printed to the log. Run `docker logs pihole | grep random` to find it.
+To set a specific password for the web interface, use the environment variable `FTLCONF_webserver_api_password`. If this variable is not detected, and you have not already set one via `pihole setpassword` / `pihole-FTL --config webserver.api.password` inside the container, then a random password will be assigned on startup, this will be printed to the log. Run `docker logs pihole | grep random password` to find it.
```
Comment (rdwebdesign) [inline README.md:87]: Changing the password is not possible when the `FTLCONF_webserver_api_password` variable is set (and any other option set by ENV VAR), but there is no explanation on the README about this ENV VARs behavior.

Comment (PromoFaux) [reply README.md:87]: Ok, I think we can mention that somewhere else on the Readme, it's a bit of a mouthful for this section

**Comment by PromoFaux** [comment]
> Test need adjustment.

Ah yes, maybe I will move back to its own function for easier testing.

File: README.md (hunk 2 above)
Comment (rdwebdesign) [reply README.md:87]: I agree. I just added the comment here because it was easier, but I think this could be in the variables section.

**Comment by rdwebdesign** [comment]
[This link](https://github.com/pi-hole/docker-pi-hole/pull/1551/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R220) in the ***Pi-hole features*** section is broken.

It is pointing to an old section:

- https://github.com/pi-hole/pi-hole/blob/master/README.md#get-help-or-connect-with-us-on-the-web
  The section **"get-help-or-connect-with-us-on-the-web"** was [removed in 2017](https://github.com/pi-hole/pi-hole/commit/a458482e51b981358d905d1c55957ffb3a2670d8#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L261).

Also, we should use a link to the repo page _(example: `https://github.com/pi-hole/pi-hole#section-ID`)_, instead of the file contents _(`https://github.com/pi-hole/pi-hole/blob/master/README.md#section-ID`)_.

File: test/tests/test_bash_functions.py [hunk 3]
```diff
@@ -31,12 +29,3 @@ def test_random_password_assigned_fresh_start(docker):
 def test_password_set_by_envvar(docker):
     func = docker.run(CMD_SETUP_WEB_PASSWORD)
     assert "Assigning password defined by Environment Variable" in func.stdout
-
-
-@pytest.mark.parametrize("test_args", ['-e "FTLCONF_ENV_ONLY=true"'])
-def test_password_envonly_true(docker):
```
Comment (PromoFaux) [inline test/tests/test_bash_functions.py:37]: adam mode can go away now @DL6ER

Comment (DL6ER) [reply test/tests/test_bash_functions.py:37]: https://github.com/pi-hole/FTL/pull/1906

**Review by rdwebdesign — APPROVED** [review]

**Pull request merged**